src/Polysemy/Final/Async.hs

```diff
@@ -18,21 +20,42 @@ import Polysemy.Final
 --
 -- /Beware/: Effects that aren't interpreted in terms of 'IO'
 -- will have local state semantics in regards to 'Async' effects
--- interpreted this way. See 'interpretFinal'.
+-- interpreted this way. See 'Final'.
 --
 -- Notably, unlike 'asyncToIO', this is not consistent with
--- 'Polysemy.State.State' unless 'Polysemy.State.runStateInIORef' is used.
+-- 'Polysemy.State.State' unless 'Polysemy.State.runStateIORef' is used.
 -- State that seems like it should be threaded globally throughout the `Async`
 -- /will not be./
 --
--- Prefer 'asyncToIO' unless its unsafe or inefficient in the context of your
--- application.
-runAsyncFinal :: Member (Final IO) r
-              => Sem (Async ': r) a
-              -> Sem r a
-runAsyncFinal = interpretFinal $ \case
+-- Prefer 'asyncToIO' or 'asyncToIOThreadless' unless these are unsafe or
+-- inefficient in the context of your application.
+--
+-- Use 'asyncToIO' or 'asyncToIOFinalGlobal' instead if you need to
+-- run pure, stateful interpreters after the interpreter for 'Async'.
+asyncToIOFinal :: Member (Final IO) r
+               => Sem (Async ': r) a
+               -> Sem r a
+asyncToIOFinal = interpretFinal $ \case
+  Async m -> do
+    ins <- getInspectorS
+    m'  <- runS m
+    liftS $ A.async (inspect ins <$> m')
+  Await a -> liftS (A.wait a)
+{-# INLINE asyncToIOFinal #-}
+
+------------------------------------------------------------------------------
+-- | 'asyncToIOFinal' implemented using 'interpretFinalGlobal'.
+--
+-- This behaves semantically very much like 'asyncToIO',
+-- but doesn't need to spin up an interpreter thread, making it more
+-- efficient (but not any more safe).
```

Comment:
Does no more safety here mean "just as likely to deadlock"?

Reply:
Basically, yeah. It inherits `https://github.com/polysemy-research/polysemy/issues/205`. I also make no guarantee that this isn't more _unsafe_. There could manifest different issues regarding `interpretFinalGlobal` which doesn't with `withLowerToIO`.